courseworks/week5/xuanzhaogao/hw5.ipynb

```diff
@@ -0,0 +1,335 @@
+   ],
+   "source": [
+    "println(sp.colptr)\n",
+    "println(sp.rowval)\n",
+    "println(sp.nzval)"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Q2(1)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "my_spv (generic function with 1 method)"
+      ]
+     },
+     "execution_count": 13,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "function my_spv(sp, v)\n",
+    "    @assert sp.n == size(v, 1)\n",
+    "    n = sp.n\n",
+    "    m = sp.m\n",
+    "    spv = zeros(Float64, m)\n",
+    "    sp_val = sp.nzval\n",
+    "    sp_row = sp.rowval\n",
+    "    for i in 1:n\n",
+    "        for j in nzrange(sp, i)\n",
+    "            k = sp_row[j]\n",
+    "            spv[k] += sp_val[j] * v[i]\n",
+    "        end\n",
+    "    end\n",
+    "    return spv\n",
+    "end"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\u001b[0m\u001b[1mTest Summary:                         | \u001b[22m\u001b[32m\u001b[1mPass  \u001b[22m\u001b[39m\u001b[36m\u001b[1mTotal  \u001b[22m\u001b[39m\u001b[0m\u001b[1mTime\u001b[22m\n",
+      "sparse matrix - vector multiplication | \u001b[32m 100  \u001b[39m\u001b[36m  100  \u001b[39m\u001b[0m0.0s\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "Test.DefaultTestSet(\"sparse matrix - vector multiplication\", Any[], 100, false, false, true, 1.678971823313075e9, 1.678971823317142e9)"
+      ]
+     },
+     "execution_count": 14,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "using SparseArrays, Test\n",
+    "\n",
+    "@testset \"sparse matrix - vector multiplication\" begin\n",
+    "\tfor k = 1:100\n",
+    "\t\tm, n = rand(1:100, 2)\n",
+    "\t\tdensity = rand()\n",
+    "\t\tsp = sprand(m, n, density)\n",
+    "\t\tv = randn(n)\n",
+    "        @test Matrix(sp) * v ≈ my_spv(sp, v)\n",
+    "\tend\n",
+    "end"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Q2(2)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "lanczos (generic function with 1 method)"
+      ]
+     },
+     "execution_count": 15,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# here I use code in class for lanczos process\n",
+    "function lanczos(A, q1::AbstractVector{T}; abstol, maxiter) where T\n",
+    "\t# normalize the input vector\n",
+    "\tq1 = normalize(q1)\n",
+    "\t# the first iteration\n",
+    "\tq = [q1]\n",
+    "\tAq1 = A * q1\n",
+    "\tα = [q1' * Aq1]\n",
```

Comment:
Also, I notice that the type stability of my lanczos code can be improved to avoid propagation of type instability. By changing the return statement to:
```julia
return SymTridiagonal(α, β), cat(q1, q[2:end]...; dims=2)
```